### PR TITLE
Adding a mini test harness.

### PR DESCRIPTION
**Pull request by galegozi**
This test harness is designed to be a lightweight test that functions like an integration test. It runs one batch/epoch just to make sure that the code runs.
NOTE: this is not designed to be used for production, this is something we would use (internally) just to test code. This would be useful when making a new piece of code (such as a loss function).

### CODE REVIEW

**Review by hickmank — CHANGES_REQUESTED** [review]
@galegozi please see my comment in the conversation. I like the idea of this work but I want to get it to be able to run in CI from a 'flit' install.

**Comment by galegozi** [comment]
Hi @hickmank ,
I understand what you want; however, this is not designed for that.
The way I think about it, there are four levels of testing:

1. Full model test: an end-to-end test where a model is trained. These runs typically take 100+ hours. These should never, NEVER be in a CICD, unless it's for a very, very good reason (such as training a baseline once per month).
2. Quick manual test: This is a full model test except it is designed to run for a shorter period of time. This is designed to resemble the full training process as closely as possible while requiring as little time as possible. _**This (and only this) is what this PR was designed for.**_
3. CICD integration test: A lighter version of the model that can be run on CPU in 1-2 minutes or less. This kind of test can be ran on CICD periodically (such as once or twice per day). Depending on the complexity of the integration tests, it may be possible and reasonable to run it every hour; however, running it once per day is likely sufficient.
    1. Depending on the complexity of these tests, it may be reasonable to configure them to run after the PR gets all approvals and the user presses the "merge" button. Please note: we should be very, very careful what we run here, so that we are not spending 1-2 hours trying to get a PR merged; therefore, only a set of a few critical integration tests should be run after the PR merge.
4. Unit tests: The quickest (and shortest) tests that are designed to test a small portion of code (such as a method). These tests should be extremely quick, and should run in about 5-10 minutes total.

I have seen what happens when this testing fails, and it's really, really bad... Given that this project is built to be a software project, it is vital to do this right, and I will not stand to have us fail at this!
Please feel free to respond to this comment (or contact me any other way) if you have questions about this.

**Comment by galegozi** [comment]
@hickmank : Do you understand my previous comment? I don't mind writing for CLI integration tests, but I recommend refactoring the harnesses to config files first (because then we can test everything, and just have a config for integration tests) :)
My point is that the harness structure makes testing a lot more complicated... What I am aiming for is eventually to have one config file that gets loaded in with all parameters... eventually I am aiming to have a config file with defaults and where you can use regex (regular expressions), or at a bare minimum an asterisk (*) to describe the experiments.

Consider this example for the channel subsampling experiments:
```json
{
    "defaults" : {
        "param": "value"
    },
    "**1": {
        "numChannels": 1
    },
    "*1*": {
        "numEpochs": 100000,
        "numBatches": 1
    }
}
```
If that doesn't make sense, the defaults are where defaults for all experiments are stored, then the second example would match 011, 021, 031, etc, the third example would match 011, 012, 013, etc. Basically you put most of your config in defaults, then the rest of the config in the experiment specific. With this infrastructure, we could have everything (loss functions, learning rate schedulers, etc) configurable from a file. It would not support hot swap, but it could support multi-component scheduled swapping (such as swapping loss functions on a schedule).

I hope that makes sense because once this behavior is achieved, then the harnesses just become a config file, and when you run start study, it will generate the slurm script and run everything automatically. I do not expect this to be ready by tomorrow, but once all this is ready, then we can just have a config.json with all relevant parameters that generates the scripts. Eventually I'd like to have a better "run on debug" option that will try to start a study on debug (if possible), otherwise add it to normal queue. That option would not bypass debug rules, but it would fully utilize the debug queue. With the "harness as JSON", that would just be one option (set useDebug=true).

**Comment by galegozi** [comment]
@hickmank Do you insist on the CI test bed for this PR to make it in? Or can we merge this in separately (this change is a completely separate test bed)?

If you insist on CI, then I will close this PR (because doing CI integration testing will take time)

File: applications/harnesses/mini-run-test/local-yoke-runner.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export PYTHONPATH=../../../src:$PYTHONPATH
```
Comment (hickmank) [inline applications/harnesses/mini-run-test/local-yoke-runner.sh:1]: I don't want to include shell scripts with python path manipulation in the repo. If `yoke` is installed properly then a user can execute `python START_study.py` from the command line.

File: applications/harnesses/mini-run-test/hyperparameters.csv
Comment (hickmank) [inline applications/harnesses/mini-run-test/hyperparameters.csv:1]: Your description of this PR is that a single batch will be run as a test. However, this `hyperparameters.csv` file has many more studies included. Also, you have one study that is labelled 926. Let's call it 1 and only have one study if this is just for testing.

**Review by hickmank — CHANGES_REQUESTED** [review]
A few minor changes. I still think a CI job that just calls `python START_study.py` in this directory with random data would be more useful but I'll leave it up to you.

File: applications/harnesses/mini-run-test/hyperparameters.csv
Comment (galegozi) [reply applications/harnesses/mini-run-test/hyperparameters.csv:1]: Sounds good, I'll make that change, thanks!

**Comment by hickmank** [comment]
> @hickmank : Do you understand my previous comment? I don't mind writing for CLI integration tests, but I recommend refactoring the harnesses to config files first (because then we can test everything, and just have a config for integration tests) :) My point is that the harness structure makes testing a lot more complicated... What I am aiming for is eventually to have one config file that gets loaded in with all parameters... eventually I am aiming to have a config file with defaults and where you can use regex (regular expressions), or at a bare minimum an asterisk (*) to describe the experiments.
> 
> Consider this example for the channel subsampling experiments:
> 
> ```json
> {
>     "defaults" : {
>         "param": "value"
>     },
>     "**1": {
>         "numChannels": 1
>     },
>     "*1*": {
>         "numEpochs": 100000,
>         "numBatches": 1
>     }
> }
> ```
> 
> If that doesn't make sense, the defaults are where defaults for all experiments are stored, then the second example would match 011, 021, 031, etc, the third example would match 011, 012, 013, etc. Basically you put most of your config in defaults, then the rest of the config in the experiment specific. With this infrastructure, we could have everything (loss functions, learning rate schedulers, etc) configurable from a file. It would not support hot swap, but it could support multi-component scheduled swapping (such as swapping loss functions on a schedule).
> 
> I hope that makes sense because once this behavior is achieved, then the harnesses just become a config file, and when you run start study, it will generate the slurm script and run everything automatically. I do not expect this to be ready by tomorrow, but once all this is ready, then we can just have a config.json with all relevant parameters that generates the scripts. Eventually I'd like to have a better "run on debug" option that will try to start a study on debug (if possible), otherwise add it to normal queue. That option would not bypass debug rules, but it would fully utilize the debug queue. With the "harness as JSON", that would just be one option (set useDebug=true).

This seems like a much bigger task than this PR. Sounds good but maybe it should be outlined in an issue first.

File: applications/harnesses/mini-run-test/local-yoke-runner.sh (hunk 1 above)
Comment (galegozi) [reply applications/harnesses/mini-run-test/local-yoke-runner.sh:1]: The point of this is to have the ability to have Yoke running from a different version (not the flit installed version). However, if you think that's a bad idea, I'll remove it, thanks!

**Comment by galegozi** [comment]
> A few minor changes. I still think a CI job that just calls `python START_study.py` in this directory with random data would be more useful but I'll leave it up to you.

That's a different kind of change. This is just a quick test using the real data to ensure that everything runs correctly.

A CPU test for CI would be much smaller!

I'll work on that and get a separate PR for a CPU CICD test, thanks!

**Comment by galegozi** [comment]
> > @hickmank : Do you understand my previous comment? I don't mind writing for CLI integration tests, but I recommend refactoring the harnesses to config files first (because then we can test everything, and just have a config for integration tests) :) My point is that the harness structure makes testing a lot more complicated... What I am aiming for is eventually to have one config file that gets loaded in with all parameters... eventually I am aiming to have a config file with defaults and where you can use regex (regular expressions), or at a bare minimum an asterisk (*) to describe the experiments.
> > Consider this example for the channel subsampling experiments:
> > ```json
> > {
> >     "defaults" : {
> >         "param": "value"
> >     },
> >     "**1": {
> >         "numChannels": 1
> >     },
> >     "*1*": {
> >         "numEpochs": 100000,
> >         "numBatches": 1
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If that doesn't make sense, the defaults are where defaults for all experiments are stored, then the second example would match 011, 021, 031, etc, the third example would match 011, 012, 013, etc. Basically you put most of your config in defaults, then the rest of the config in the experiment specific. With this infrastructure, we could have everything (loss functions, learning rate schedulers, etc) configurable from a file. It would not support hot swap, but it could support multi-component scheduled swapping (such as swapping loss functions on a schedule).
> > I hope that makes sense because once this behavior is achieved, then the harnesses just become a config file, and when you run start study, it will generate the slurm script and run everything automatically. I do not expect this to be ready by tomorrow, but once all this is ready, then we can just have a config.json with all relevant parameters that generates the scripts. Eventually I'd like to have a better "run on debug" option that will try to start a study on debug (if possible), otherwise add it to normal queue. That option would not bypass debug rules, but it would fully utilize the debug queue. With the "harness as JSON", that would just be one option (set useDebug=true).
> 
> This seems like a much bigger task than this PR. Sounds good but maybe it should be outlined in an issue first.

Of course this is not an issue for one PR! but once that's done (as in "harness as JSON"), then testing would be much easier and more comprehensive...
Imagine there being unit tests for the harnesses!

**Comment by galegozi** [comment]
> > @hickmank Do you insist on the CI test bed for this PR to make it in? Or can we merge this in separately (this change is a completely separate test bed)?
> > If you insist on CI, then I will close this PR (because doing CI integration testing will take time)
> 
> I made a few comments in the review with minor changes requested. Notably, there does not seem to be a need for the extra shell script.
> 
> It seems like it should be simple to have a CI job that just calls `python START_study.py` in this directory after a `flit install` and passes if one cycle of training runs. What am I missing?

The script handles multiple versions of Yoke. If this behavior does not need to be supported, then there is no need for the script.

File: applications/harnesses/mini-run-test/training_START.slurm [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# `gpu40only`.
+
+#SBATCH --job-name=loderunner_chfe_study<studyIDX>
+#SBATCH --account=w25_artimis_g
+#SBATCH --nodes=1
+#SBATCH --ntasks-per-node=64
+#SBATCH --partition=gpu_debug
+#SBATCH --reservation=gpu_debug
+#SBATCH --constraint=gpu40
+#SBATCH --time=02:00:00
```
Comment (galegozi) [inline applications/harnesses/mini-run-test/training_START.slurm:21]: shorten this

File: applications/harnesses/mini-run-test/training_input.tmpl [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+--batch_size
+<BATCH_SIZE>
+--num_workers
+<NUM_WORKERS>
+--prefetch_factor
+<PREFETCH_FACTOR>
+--total_epochs
+1
+--cycle_epochs
+10
```
Comment (galegozi) [inline applications/harnesses/mini-run-test/training_input.tmpl:32]: run cycle epochs = 2

File: applications/harnesses/mini-run-test/training_input.tmpl [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+--val_rcrd_filename
+./validation_study<studyIDX>_epoch<epochIDX>.csv
+--batch_size
+<BATCH_SIZE>
+--num_workers
+<NUM_WORKERS>
+--prefetch_factor
+<PREFETCH_FACTOR>
+--total_epochs
+1
```
Comment (galegozi) [inline applications/harnesses/mini-run-test/training_input.tmpl:30]: total epochs = 4

File: applications/harnesses/mini-run-test/training_input.tmpl [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+--cycle_epochs
+10
+--train_batches
+<NTRN_BATCH>
+--val_batches
+<NVAL_BATCH>
+--TRAIN_PER_VAL
+5
```
Comment (galegozi) [inline applications/harnesses/mini-run-test/training_input.tmpl:38]: set it to 2

File: applications/harnesses/mini-run-test/hyperparameters.csv [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+#605,torch_ch_gpu_241112,5,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#610,torch_ch_gpu_241112,6,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#611,torch_ch_gpu_241112,4,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#620,torch_ch_gpu_241112,8,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#621,torch_ch_gpu_241112,1,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#922,torch_ch_gpu_241112,2,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#923,torch_ch_gpu_241112,3,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#924,torch_ch_gpu_241112,4,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+#925,torch_ch_gpu_241112,5,128,1,1,3,1,8,2,64,1000,500,1.00E-04,110,0.5,train_density_LodeRunner.py
+926,torch_ch_gpu_241112,6,128,1,1,3,1,8,2,64,1,1,1.00E-04,110,0.5,train_density_LodeRunner.py
```
Comment (galegozi) [inline applications/harnesses/mini-run-test/hyperparameters.csv:27]: study ID 1, batch size = 4

**Comment by galegozi** [comment]
ADD A README

**Comment by galegozi** [comment]
@hickmank : I made the requested changes.
Could you please re-review this PR? thanks!

File: applications/harnesses/mini-run-test/training_START.slurm [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+conda activate /usr/projects/artimis/conda/envs/torch_ch_gpu_241112/
+
+# Currently need to set an environment variable for MKL
+# threading. Believed to be related to Numpy.
+export MKL_SERVICE_FORCE_INTEL=TRUE
+
+# Get start time
+export date00=`date`
+
+# # Start GPU monitoring
```
Comment (hickmank) [inline applications/harnesses/mini-run-test/training_START.slurm:44]: It is not important but I'd get rid of this old commented code. It is not used anymore and I can't foresee a time when we'd want it.

**Review by hickmank — APPROVED** [review]
The slurm file and template code be cleaned a little more. Other than that, looks good!

File: applications/harnesses/mini-run-test/training_START.slurm (hunk 7 above)
Comment (galegozi) [reply applications/harnesses/mini-run-test/training_START.slurm:44]: I will resolve this comment for now, given that the plan is to "de-slurm" the repo :)
